internal/jobs/iocp.go

```diff
@@ -0,0 +1,58 @@
+package jobs
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	ioInitOnce sync.Once
+	// Global iocp handle that will be re-used for every job object
+	ioCompletionPort windows.Handle
+	// Mapping of job handle to channel that that job object owns to be able to receive
+	// notifications.
+	jobMap = make(map[uint32]chan IOCPNotif)
+)
+
+// IOCPNotif represents a job object specific IOCP message.
+// Code will hold the value of the job object specific message that was received.
+// Key will hold the job object handle value so the job the message is for can be determined.
+// Err holds any error returned from polling the IOCP.
+type IOCPNotif struct {
+	Code uint32
+	Key  uint32
+	Err  error
+}
+
+// ioCompletionProcessor processes completed async IOs forever
+func pollIOCP(iocpHandle windows.Handle) {
+	var (
+		overlapped uintptr
+		code       uint32
+		key        uint32
+	)
+
+	for {
+		err := windows.GetQueuedCompletionStatus(iocpHandle, &code, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE)
+		if ch, ok := jobMap[key]; ok {
+			ch <- IOCPNotif{
+				Code: code,
+				Key:  key,
+				Err:  err,
+			}
+		}
+	}
+}
+
+// Assigns an IO completion port to get notified of events for the registered job
+// object.
+func attachIOCP(job windows.Handle, iocp windows.Handle) (int, error) {
```

Comment:
Seems like there is no reason for this function to return an `int`, as it's never used.